csrc/multidevice/collective.cpp

```diff
@@ -0,0 +1,199 @@
+  TORCH_INTERNAL_ASSERT(
+      std::find(
+          self.params().team.begin(),
+          self.params().team.end(),
+          comm.deviceId()) != self.params().team.end(),
+      "current device index ",
+      comm.deviceId(),
+      " must be present in the collective's team");
+}
+
+} // namespace
+
+Collective::Collective(CollectiveParams params, std::string name, bool has_root)
+    : params_(std::move(params)),
+      collective_type_(std::move(name)),
+      has_root_(has_root) {
+  assertBuffersHaveSameSize(params_.src_bufs, params_.dst_bufs);
+  TORCH_INTERNAL_ASSERT(
+      std::unique(params_.team.begin(), params_.team.end()) ==
+          params_.team.end(),
+      "the collective must not involve the same device more than once");
+  TORCH_INTERNAL_ASSERT(
+      params_.team.size() > 1, "the team size must be greater than 1");
+  if (has_root_) {
+    auto it = std::find(params_.team.begin(), params_.team.end(), params_.root);
+    TORCH_INTERNAL_ASSERT(
+        it != params_.team.end(),
+        "root (device ",
+        params_.root,
+        ") must be present in the collective's team");
+    // pytorch's process group expects the root to be specified
+    // by its relative rank, i.e., the device index's offset within the team
+    root_rank_ = std::distance(params_.team.begin(), it);
+  }
+}
+
+std::string Collective::toString(int indent) const {
+  std::stringstream ss;
+  std::string ext_indent(" ", indent);
+  std::string indent1 = ext_indent + "  ";
+  std::string indent2 = ext_indent + "    ";
+
+  ss << ext_indent << "Collective " << collective_type_ << ": {\n";
+
+  if (has_root_) {
+    ss << indent1 << "root: " << params_.root << ",\n";
+  }
+  ss << indent1 << "team: {";
+  for (auto r : params_.team) {
+    ss << r << ", ";
+  }
+  ss << indent1 << "}\n";
+  ss << indent1 << "src_bufs: {";
+  for (auto& t : params_.src_bufs) {
+    ss << "\n" << t;
+  }
+  ss << "\n" << indent1 << "}\n";
+  ss << ext_indent << "}";
+
+  return ss.str();
+}
+
+Broadcast::Broadcast(CollectiveParams params)
+    : Collective(params, "broadcast") {}
+
+c10::intrusive_ptr<c10d::Work> Broadcast::post(Communicator& comm) {
+  post_common(*this, comm);
+
+  if (comm.deviceId() == params_.root) {
+    assertBufferCount(params_.src_bufs, 1);
+    assertBufferCount(params_.dst_bufs, 0);
+  } else {
+    assertBufferCount(params_.src_bufs, 0);
+    assertBufferCount(params_.dst_bufs, 1);
+  }
+
+  return comm.getTeam(params_.team)
+      ->broadcast(
+          comm.deviceId() == params_.root ? params_.src_bufs : params_.dst_bufs,
+          {.rootRank = root_rank_});
+}
+
+Gather::Gather(CollectiveParams params) : Collective(params, "gather") {
+  assertBufferCount(params_.src_bufs, 1);
+}
+
+c10::intrusive_ptr<c10d::Work> Gather::post(Communicator& comm) {
+  post_common(*this, comm);
+
+  if (comm.deviceId() == params_.root) {
+    assertBufferCount(params_.dst_bufs, params_.team.size());
+    buf_list_ = setBufList(params_.dst_bufs);
```

Comment:
Why can't this be just a local variable? Does its lifetime need to be persistent?

Reply:
Not sure what you mean. For example, the gather is posted as:

```
comm.getBackendForTeam(params_.team)
      ->gather(buf_list_, params_.src_bufs, {.rootRank = root_relative_index_});
```

Here, after returning from this gather call, it's not guaranteed that the communication is completed, so we need to make sure the destination buffers are still alive. However, do we still need to make sure `buf_list_` itself alive?

Is this the actual `gather` function we are using?

https://github.com/pytorch/pytorch/blob/e1aba2c8c39d817da391a75aa7ce6dd590f93755/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L3090C26-L3173

If so, it seems the output list is copied into another vector, `outputs`, so it seems we don't need to retain `buf_list_`. One thing not clear for me is why the `outputTensors` parameter needs to be a non-const reference, even though it doesn't seem to be modified.

> Recall that we need to be able to post the collective multiple times

Do we have tests?

Reply:
> we need to make sure the destination buffers are still alive. However, do we still need to make sure `buf_list_` itself alive?

Correct, we need the destination buffer to stay alive but not necessarily `buf_list_`. I have no problem changing, however, I am not sure to understand what you suggest. Could you clarify how it should be done?

> Is this the actual `gather` function we are using?
> 
> https://github.com/pytorch/pytorch/blob/e1aba2c8c39d817da391a75aa7ce6dd590f93755/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L3090C26-L3173

yes



> Do we have tests?

good point, let me add tests for that

Reply:
> why the outputTensors parameter needs to be a non-const reference

I don't know...

Reply:
> Correct, we need the destination buffer to stay alive but not necessarily buf_list_. I have no problem changing, however, I am not sure to understand what you suggest. Could you clarify how it should be done?

Can we just make `buf_list_` a local variable? It's just, the simpler, the better.

Reply:
> good point, let me add tests for that

Is this done?

Reply:
> Can we just make `buf_list_` a local variable? It's just, the simpler, the better

Can you be more precise ? What I claim is that the current way is the simplest.

Do you agree that, for performance reason, we need to "`std::move`" the dst_bufs to buf_lists_? If so, making buf_list a local variable will make `posts` look like:
```
c10::intrusive_ptr<c10d::Work> Gather::post(Communicator& comm) {
[...]
std::vector<std::vector<at::Tensor>> buf_list = {std::move(params_.dst_bufs)}
auto work = comm.getBackendForTeam(params_.team)
      ->gather(buf_list, params_.src_bufs, {.rootRank = root_relative_index_});
params_.dst_bufs = std::move(buf_list.back());
}
```
I think the current solution is simpler.
Do you have something else in mind ?

Reply:
> > good point, let me add tests for that
> 
> Is this done?

It is, in the last commits. I'm just hardening again the tests now by making buffers non identical throughout iterations